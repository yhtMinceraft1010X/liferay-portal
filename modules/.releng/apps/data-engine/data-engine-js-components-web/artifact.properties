artifact.git.id=b80ee6bc52c0532e058704f1fefbd2d16f6d6e2a
artifact.jsdoc.url=https://repository-cdn.liferay.com/nexus/content/repositories/liferay-public-releases/com/liferay/com.liferay.data.engine.js.components.web/1.0.60/com.liferay.data.engine.js.components.web-1.0.60-jsdoc.jar
artifact.sources-commercial.url=https://repository-cdn.liferay.com/nexus/content/repositories/liferay-public-releases/com/liferay/com.liferay.data.engine.js.components.web/1.0.60/com.liferay.data.engine.js.components.web-1.0.60-sources-commercial.jar
artifact.sources.url=https://repository-cdn.liferay.com/nexus/content/repositories/liferay-public-releases/com/liferay/com.liferay.data.engine.js.components.web/1.0.60/com.liferay.data.engine.js.components.web-1.0.60-sources.jar
artifact.url=https://repository-cdn.liferay.com/nexus/content/repositories/liferay-public-releases/com/liferay/com.liferay.data.engine.js.components.web/1.0.60/com.liferay.data.engine.js.components.web-1.0.60.jar