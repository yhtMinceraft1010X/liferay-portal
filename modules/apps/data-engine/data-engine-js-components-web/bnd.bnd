Bundle-Name: Liferay Data Engine JS Components Web
Bundle-SymbolicName: com.liferay.data.engine.js.components.web
Bundle-Version: 1.0.61
Web-ContextPath: /data-engine-js-components-web