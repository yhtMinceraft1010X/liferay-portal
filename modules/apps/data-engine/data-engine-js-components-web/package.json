{
	"dependencies": {
		"dom-align": "1.10.4",
		"moment": "^2.22.2"
	},
	"jest": {
		"collectCoverageFrom": [
			"src/main/resources/META-INF/resources/js/**/*.es.js"
		],
		"modulePathIgnorePatterns": [
			"/__mock__/",
			"/__utils__/"
		],
		"setupFiles": [
			"<rootDir>/jest-setup.config.js"
		],
		"snapshotSerializers": [
			"metal-jest-serializer"
		],
		"testPathIgnorePatterns": [
			"<rootDir>/test/js/[^/]+\\.js"
		]
	},
	"main": "js/index.es.js",
	"name": "data-engine-js-components-web",
	"private": true,
	"scripts": {
		"build": "liferay-npm-scripts build",
		"checkFormat": "liferay-npm-scripts check",
		"format": "liferay-npm-scripts fix",
		"test": "metalsoy -s \"test/**/*.soy\" -d \"test\" && liferay-npm-scripts test"
	},
	"version": "1.0.61"
}
