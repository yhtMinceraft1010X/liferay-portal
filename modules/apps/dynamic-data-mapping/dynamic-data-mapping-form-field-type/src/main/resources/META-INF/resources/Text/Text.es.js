/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import ClayAutocomplete from '@clayui/autocomplete';
import ClayDropDown from '@clayui/drop-down';
import {ClayInput} from '@clayui/form';
import {usePrevious} from '@liferay/frontend-js-react-web';
import {normalizeFieldName} from 'data-engine-js-components-web';
import React, {useCallback, useEffect, useMemo, useRef, useState} from 'react';

import {FieldBase} from '../FieldBase/ReactFieldBase.es';
import {useSyncValue} from '../hooks/useSyncValue.es';
import withConfirmationField from '../util/withConfirmationField.es';

const CounterContainer = ({
	counter,
	displayErrors,
	error,
	maxLength,
	setError,
	showCounter,
	valid,
}) => {
	if (
		!showCounter ||
		(displayErrors === true && !valid && Object.keys(error).length === 0)
	) {
		return null;
	}

	const message = Liferay.Util.sub(
		Liferay.Language.get('x-of-x-characters'),
		counter,
		maxLength
	);

	if (counter > maxLength) {
		if (error.errorMessage !== message) {
			setError({
				displayErrors: true,
				errorMessage: message,
				valid: false,
			});
		}

		return null;
	}

	if (error.displayErrors) {
		setError({});
	}

	return (
		<span aria-hidden="true" className="form-text">
			{message}
		</span>
	);
};

const Text = ({
	defaultLanguageId,
	disabled,
	displayErrors,
	editingLanguageId,
	error,
	fieldName,
	id,
	invalidCharacters,
	localizable,
	localizedValue,
	maxLength,
	name,
	normalizeField,
	onBlur,
	onChange,
	onFocus,
	placeholder,
	setError,
	shouldUpdateValue,
	showCounter,
	syncDelay,
	valid,
	value: initialValue,
}) => {
	const [value, setValue] = useSyncValue(
		initialValue,
		syncDelay,
		editingLanguageId
	);

	const inputRef = useRef(null);

	const prevEditingLanguageId = usePrevious(editingLanguageId);

	useEffect(() => {
		if (prevEditingLanguageId !== editingLanguageId && localizable) {
			const newValue =
				localizedValue[editingLanguageId] !== undefined
					? localizedValue[editingLanguageId]
					: localizedValue[defaultLanguageId];
			setValue(newValue);
		}
	}, [
		defaultLanguageId,
		editingLanguageId,
		localizable,
		localizedValue,
		prevEditingLanguageId,
		setValue,
	]);

	useEffect(() => {
		if (
			normalizeField &&
			inputRef.current &&
			inputRef.current.value !== initialValue &&
			(inputRef.current.value === '' || shouldUpdateValue)
		) {
			setValue(initialValue);
			onChange({target: {value: initialValue}});
		}
	}, [
		initialValue,
		inputRef,
		fieldName,
		normalizeField,
		onChange,
		setValue,
		shouldUpdateValue,
	]);

	return (
		<>
			<ClayInput
				className="ddm-field-text"
				dir={Liferay.Language.direction[editingLanguageId]}
				disabled={disabled}
				id={id}
				lang={editingLanguageId}
				maxLength={showCounter ? '' : maxLength}
				name={name}
				onBlur={(event) => {
					if (normalizeField) {
						onBlur({target: {value: initialValue}});
					}
					else {
						onBlur(event);
					}
				}}
				onChange={(event) => {
					const {value} = event.target;

					if (normalizeField) {
						event.target.value = normalizeFieldName(value);
					}
					else if (invalidCharacters) {
						const regex = new RegExp(invalidCharacters, 'g');

						event.target.value = value.replace(regex, '');
					}
					setValue(event.target.value);
					onChange(event);
				}}
				onFocus={onFocus}
				placeholder={placeholder}
				ref={inputRef}
				type="text"
				value={value}
			/>

			<CounterContainer
				counter={value?.length}
				displayErrors={displayErrors}
				error={error}
				maxLength={maxLength}
				setError={setError}
				showCounter={showCounter}
				valid={valid}
			/>
		</>
	);
};

const Textarea = ({
	disabled,
	displayErrors,
	editingLanguageId,
	error,
	id,
	maxLength,
	name,
	onBlur,
	onChange,
	onFocus,
	placeholder,
	setError,
	showCounter,
	syncDelay,
	valid,
	value: initialValue,
}) => {
	const [value, setValue] = useSyncValue(initialValue, syncDelay);

	return (
		<>
			<textarea
				className="ddm-field-text form-control"
				dir={Liferay.Language.direction[editingLanguageId]}
				disabled={disabled}
				id={id}
				lang={editingLanguageId}
				name={name}
				onBlur={onBlur}
				onChange={(event) => {
					setValue(event.target.value);
					onChange(event);
				}}
				onFocus={onFocus}
				placeholder={placeholder}
				style={disabled ? {resize: 'none'} : null}
				type="text"
				value={value}
			/>

			<CounterContainer
				counter={value?.length}
				displayErrors={displayErrors}
				error={error}
				maxLength={maxLength}
				setError={setError}
				showCounter={showCounter}
				valid={valid}
			/>
		</>
	);
};

const Autocomplete = ({
	disabled,
	editingLanguageId,
	id,
	name,
	onBlur,
	onChange,
	onFocus,
	options,
	placeholder,
	syncDelay,
	value: initialValue,
}) => {
	const [selectedItem, setSelectedItem] = useState(false);
	const [value, setValue] = useSyncValue(initialValue, syncDelay);
	const [visible, setVisible] = useState(false);
	const inputRef = useRef(null);
	const itemListRef = useRef(null);

	const escapeChars = (string) =>
		string.replace(/[.*+\-?^${}()|[\]\\]/g, '\\$&');

	const filteredItems = options.filter(
		(item) => item && item.match(escapeChars(value))
	);

	const isValidItem = useCallback(() => {
		return (
			!selectedItem &&
			filteredItems.length > 1 &&
			!filteredItems.includes(value)
		);
	}, [filteredItems, selectedItem, value]);

	useEffect(() => {
		const ddmPageContainerLayout = inputRef.current.closest(
			'.ddm-page-container-layout'
		);

		if (
			!isValidItem() &&
			ddmPageContainerLayout &&
			ddmPageContainerLayout.classList.contains('hide')
		) {
			setVisible(false);
		}
	}, [filteredItems, isValidItem, value, selectedItem]);

	const handleFocus = (event, direction) => {
		const target = event.target;
		const focusabledElements = event.currentTarget.querySelectorAll(
			'button'
		);
		const targetIndex = [...focusabledElements].findIndex(
			(current) => current === target
		);

		let nextElement;

		if (direction) {
			nextElement = focusabledElements[targetIndex - 1];
		}
		else {
			nextElement = focusabledElements[targetIndex + 1];
		}

		if (nextElement) {
			event.preventDefault();
			event.stopPropagation();
			nextElement.focus();
		}
		else if (targetIndex === 0 && direction) {
			event.preventDefault();
			event.stopPropagation();
			inputRef.current.focus();
		}
	};

	return (
		<ClayAutocomplete>
			<ClayAutocomplete.Input
				dir={Liferay.Language.direction[editingLanguageId]}
				disabled={disabled}
				id={id}
				lang={editingLanguageId}
				name={name}
				onBlur={onBlur}
				onChange={(event) => {
					setValue(event.target.value);
					setVisible(!!event.target.value);
					setSelectedItem(false);
					onChange(event);
				}}
				onFocus={(event) => {
					if (isValidItem() && event.target.value) {
						setVisible(true);
					}

					onFocus(event);
				}}
				onKeyDown={(event) => {
					if (
						(event.key === 'Tab' || event.key === 'ArrowDown') &&
						!event.shiftKey &&
						filteredItems.length > 0 &&
						visible
					) {
						event.preventDefault();
						event.stopPropagation();

						const firstElement = itemListRef.current.querySelector(
							'button'
						);
						firstElement.focus();
					}
				}}
				placeholder={placeholder}
				ref={inputRef}
				value={value}
			/>

			<ClayAutocomplete.DropDown
				active={visible && !disabled}
				onSetActive={setVisible}
			>
				<ul
					className="list-unstyled"
					onKeyDown={(event) => {
						switch (event.key) {
							case 'ArrowDown':
								handleFocus(event, false);
								break;
							case 'ArrowUp':
								handleFocus(event, true);
								break;
							case 'Tab':
								handleFocus(event, event.shiftKey);
								break;
							default:
								break;
						}
					}}
					ref={itemListRef}
				>
					{filteredItems.length === 0 && (
						<ClayDropDown.Item className="disabled">
							{Liferay.Language.get('no-results-were-found')}
						</ClayDropDown.Item>
					)}

					{filteredItems.map((label, index) => (
						<ClayAutocomplete.Item
							key={index}
							match={value}
							onClick={() => {
								setValue(label);
								setVisible(false);
								setSelectedItem(true);
								onChange({target: {value: label}});
							}}
							value={label}
						/>
					))}
				</ul>
			</ClayAutocomplete.DropDown>
		</ClayAutocomplete>
	);
};

const DISPLAY_STYLE = {
	autocomplete: Autocomplete,
	multiline: Textarea,
	singleline: Text,
};

const Main = ({
	autocomplete,
	autocompleteEnabled,
	defaultLanguageId,
	displayErrors,
	displayStyle = 'singleline',
	showCounter,
	fieldName,
	id,
	invalidCharacters = '',
	locale,
	localizable,
	localizedValue = {},
	maxLength,
	name,
	normalizeField = false,
	onBlur,
	onChange,
	onFocus,
	options = [],
	placeholder,
	predefinedValue = '',
	readOnly,
	shouldUpdateValue = false,
	syncDelay = true,
	valid,
	value,
	...otherProps
}) => {
	const optionsMemo = useMemo(() => options.map((option) => option.label), [
		options,
	]);

	const [error, setError] = useState({});

	const Component =
		DISPLAY_STYLE[
			autocomplete || autocompleteEnabled
				? 'autocomplete'
				: displayStyle ?? `singleline`
		];

	return (
		<FieldBase
			{...otherProps}
			{...error}
			displayErrors={error.displayErrors ?? displayErrors}
			fieldName={fieldName}
			id={id}
			localizedValue={localizedValue}
			name={name}
			readOnly={readOnly}
			valid={error.valid ?? valid}
		>
			<Component
				defaultLanguageId={defaultLanguageId}
				disabled={readOnly}
				displayErrors={displayErrors}
				editingLanguageId={locale}
				error={error}
				fieldName={fieldName}
				id={id ?? name}
				invalidCharacters={invalidCharacters}
				localizable={localizable}
				localizedValue={localizedValue}
				maxLength={maxLength}
				name={name}
				normalizeField={normalizeField}
				onBlur={onBlur}
				onChange={onChange}
				onFocus={onFocus}
				options={optionsMemo}
				placeholder={placeholder}
				setError={setError}
				shouldUpdateValue={shouldUpdateValue}
				showCounter={showCounter}
				syncDelay={syncDelay}
				valid={valid}
				value={value ? value : predefinedValue}
			/>
		</FieldBase>
	);
};

Main.displayName = 'Text';

export default withConfirmationField(Main);
