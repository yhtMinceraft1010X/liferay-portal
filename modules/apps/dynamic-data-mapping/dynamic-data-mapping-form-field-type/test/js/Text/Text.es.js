/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import '@testing-library/jest-dom/extend-expect';
import {act, cleanup, fireEvent, render, screen} from '@testing-library/react';
import {PageProvider} from 'data-engine-js-components-web';
import React from 'react';

import Text from '../../../src/main/resources/META-INF/resources/Text/Text.es';

const globalLanguageDirection = Liferay.Language.direction;

const spritemap = 'icons.svg';

const defaultTextConfig = {
	name: 'textField',
	spritemap,
};

const TextWithProvider = (props) => (
	<PageProvider value={{editingLanguageId: 'en_US'}}>
		<Text {...props} />
	</PageProvider>
);

describe('Field Text', () => {
	// eslint-disable-next-line no-console
	const originalWarn = console.warn;

	beforeAll(() => {
		// eslint-disable-next-line no-console
		console.warn = (...args) => {
			if (/DataProvider: Trying/.test(args[0])) {
				return;
			}
			originalWarn.call(console, ...args);
		};

		Liferay.Language.direction = {
			en_US: 'rtl',
		};
	});

	afterAll(() => {
		// eslint-disable-next-line no-console
		console.warn = originalWarn;

		Liferay.Language.direction = globalLanguageDirection;
	});

	afterEach(cleanup);

	beforeEach(() => {
		jest.useFakeTimers();
		fetch.mockResponseOnce(JSON.stringify({}));
	});

	it('is not readOnly', () => {
		const {container} = render(
			<TextWithProvider {...defaultTextConfig} readOnly={false} />
		);

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});

	it('is readOnly', () => {
		const {container} = render(
			<TextWithProvider {...defaultTextConfig} readOnly={true} />
		);

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});

	it('has a helptext', () => {
		const {container} = render(
			<TextWithProvider {...defaultTextConfig} tip="Type something" />
		);

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});

	it('has an id', () => {
		const {container} = render(
			<TextWithProvider {...defaultTextConfig} id="Id" />
		);

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});

	it('has a label', () => {
		const {container} = render(
			<TextWithProvider {...defaultTextConfig} label="label" />
		);

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});

	it('has a placeholder', () => {
		const {container} = render(
			<TextWithProvider
				{...defaultTextConfig}
				placeholder="Placeholder"
			/>
		);

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});

	it('hides autocomplete dropdown menu when container layout is hidden', () => {
		const props = {
			autocomplete: true,
			options: [
				{label: 'Option 1', value: 'Option1'},
				{label: 'Option 2', value: 'Option2'},
			],
			value: 'Option',
			...defaultTextConfig,
		};

		render(
			<div className="ddm-page-container-layout hide">
				<TextWithProvider {...props} />
			</div>
		);

		act(() => {
			jest.runAllTimers();
		});

		const autocompleteDropdownMenu = document.body.querySelector(
			'.autocomplete-dropdown-menu'
		);

		const classList = autocompleteDropdownMenu.classList;

		expect(classList.contains('show')).toBeFalsy();
	});

	it('is not required', () => {
		const {container} = render(
			<TextWithProvider {...defaultTextConfig} required={false} />
		);

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});

	it('renders autocomplete dropdown menu', () => {
		const onChange = jest.fn();

		const props = {
			autocomplete: true,
			options: [
				{label: 'Option 1', value: 'Option1'},
				{label: 'Option 2', value: 'Option2'},
			],
			value: '',
			...defaultTextConfig,
		};

		const {container} = render(
			<div className="ddm-page-container-layout">
				<TextWithProvider {...props} key="input" onChange={onChange} />
			</div>
		);

		const input = container.querySelector('input');

		fireEvent.change(input, {
			target: {
				value: 'Option',
			},
		});

		act(() => {
			jest.runAllTimers();
		});

		const autocompleteDropdownMenu = document.querySelector(
			'.autocomplete-dropdown-menu'
		);

		expect(
			autocompleteDropdownMenu.classList.contains('show')
		).toBeTruthy();
	});

	it('hides autocomplete dropdown menu when input is empty', () => {
		const onChange = jest.fn();

		const props = {
			autocomplete: true,
			options: [
				{label: 'Option 1', value: 'Option1'},
				{label: 'Option 2', value: 'Option2'},
			],
			value: 'Option',
			...defaultTextConfig,
		};

		const {container} = render(
			<div className="ddm-page-container-layout">
				<TextWithProvider {...props} key="input" onChange={onChange} />
			</div>
		);

		const input = container.querySelector('input');

		fireEvent.change(input, {
			target: {
				value: '',
			},
		});

		act(() => {
			jest.runAllTimers();
		});

		const autocompleteDropdownMenu = document.querySelector(
			'.autocomplete-dropdown-menu'
		);

		expect(autocompleteDropdownMenu.classList.contains('show')).toBeFalsy();
	});

	it('hides autocomplete dropdown menu when focus is changed', () => {
		const onChange = jest.fn();

		const props = {
			autocomplete: true,
			options: [
				{label: 'Option 1', value: 'Option1'},
				{label: 'Option 2', value: 'Option2'},
			],
			value: '',
			...defaultTextConfig,
		};

		const {container} = render(
			<div className="ddm-page-container-layout">
				<TextWithProvider {...props} key="input" onChange={onChange} />
			</div>
		);

		const input = container.querySelector('input');

		fireEvent.change(input, {
			target: {
				value: 'Option',
			},
		});

		act(() => {
			jest.runAllTimers();
		});

		const autocompleteDropdownMenu = document.querySelector(
			'.autocomplete-dropdown-menu'
		);

		expect(
			autocompleteDropdownMenu.classList.contains('show')
		).toBeTruthy();

		const body = document.body;

		fireEvent.mouseDown(body);

		act(() => {
			jest.runAllTimers();
		});

		expect(autocompleteDropdownMenu.classList.contains('show')).toBeFalsy();
	});

	it('renders Label if showLabel is true', () => {
		const {container} = render(
			<TextWithProvider {...defaultTextConfig} label="text" showLabel />
		);

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});

	it('has a value', () => {
		const {container} = render(
			<TextWithProvider {...defaultTextConfig} value="value" />
		);

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});

	it('emits a field edit with correct parameters', () => {
		const onChange = jest.fn();

		const {container} = render(
			<TextWithProvider
				{...defaultTextConfig}
				key="input"
				onChange={onChange}
			/>
		);

		const input = container.querySelector('input');

		fireEvent.change(input, {
			target: {
				value: 'test',
			},
		});

		act(() => {
			jest.runAllTimers();
		});

		expect(onChange).toHaveBeenCalled();
	});

	it('normalizes the field if it contains invalid characters', () => {
		const onChange = jest.fn();

		const {container} = render(
			<TextWithProvider
				{...defaultTextConfig}
				key="input"
				normalizeField={true}
				onChange={onChange}
			/>
		);

		const input = container.querySelector('input');

		fireEvent.change(input, {
			target: {
				value: 'Field¿êReference',
			},
		});

		act(() => {
			jest.runAllTimers();
		});

		expect(input.value).toEqual('FieldReference');
	});

	it('normalizes the value of the field if it contains invalid characters', () => {
		const onChange = jest.fn();

		const {container} = render(
			<TextWithProvider
				{...defaultTextConfig}
				invalidCharacters="[1-8]"
				key="input"
				onChange={onChange}
			/>
		);

		const input = container.querySelector('input');

		fireEvent.change(input, {
			target: {
				value: '+9 (129) 993-9999',
			},
		});

		act(() => {
			jest.runAllTimers();
		});

		expect(input.value).toEqual('+9 (9) 99-9999');
	});

	it('renders a counter when show counter is true, there is a maxLength and value is empty', () => {
		const {getByText} = render(
			<TextWithProvider
				{...defaultTextConfig}
				maxLength={10}
				showCounter={true}
				valid={true}
				value=""
			/>
		);

		expect(getByText('0/10 characters')).toBeInTheDocument();
	});

	it('renders a counter when show counter is true, there is a maxLength and value is different from empty', () => {
		const {getByText} = render(
			<TextWithProvider
				{...defaultTextConfig}
				maxLength={10}
				showCounter={true}
				valid={true}
				value="test"
			/>
		);

		expect(getByText('4/10 characters')).toBeInTheDocument();
	});

	it('does not render a counter when show counter is false, there is a maxLength and value is different from empty', () => {
		const {queryByText} = render(
			<TextWithProvider
				{...defaultTextConfig}
				maxLength={10}
				value="test"
			/>
		);

		expect(queryByText('4/10 characters')).not.toBeInTheDocument();
	});

	it('renders a counter when show counter is true, there is a maxLength and value length is greater than the maximum lenght', () => {
		render(
			<TextWithProvider
				{...defaultTextConfig}
				maxLength={2}
				showCounter={true}
				valid={true}
				value="test"
			/>
		);
		const error = screen.queryAllByText('4/2 characters')[0];
		expect(error).toHaveClass('form-feedback-item');
	});

	describe('Confirmation Field', () => {
		it('does not show the confirmation field', () => {
			render(<TextWithProvider {...defaultTextConfig} />);

			const confirmationField = document.getElementById(
				'textFieldconfirmationField_fieldDetails'
			);

			expect(confirmationField).toBeNull();
		});

		it('shows the confirmation field if the requireConfirmation property is enabled', () => {
			const {container} = render(
				<TextWithProvider
					{...defaultTextConfig}
					direction="horizontal"
					requireConfirmation={true}
				/>
			);

			const confirmationField = document.getElementById(
				'textFieldconfirmationField'
			);

			expect(confirmationField).not.toBeNull();

			expect(container.firstChild).toHaveClass('row');

			expect(
				container.firstChild.querySelector('.col-md-6')
			).not.toBeNull();
		});

		it('shows the confirmation field in vertical mode', () => {
			const {container} = render(
				<TextWithProvider
					{...defaultTextConfig}
					direction="vertical"
					requireConfirmation={true}
				/>
			);

			expect(
				container.firstChild.querySelector('.col-md-12')
			).not.toBeNull();
		});
	});
});
