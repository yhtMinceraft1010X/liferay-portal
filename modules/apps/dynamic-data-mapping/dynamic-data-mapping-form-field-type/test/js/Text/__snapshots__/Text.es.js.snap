// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field Text has a helptext 1`] = `
<div>
  <div
    aria-labelledby="textField_fieldDetails"
    class="form-group hide"
    data-field-name="textField"
    tabindex="0"
  >
    <input
      class="form-control ddm-field-text"
      id="textField"
      name="textField"
      type="text"
      value=""
    />
    <input
      name="textField_edited"
      type="hidden"
      value="false"
    />
    <div
      aria-hidden="true"
      class="form-feedback-group lfr-de__field-feedback"
    >
      <div>
        Type something
      </div>
    </div>
    <span
      class="sr-only"
      id="textField_fieldDetails"
    >
      Type something
    </span>
  </div>
</div>
`;

exports[`Field Text has a label 1`] = `
<div>
  <div
    class="form-group hide"
    data-field-name="textField"
  >
    <label
      aria-labelledby="textField_fieldDetails"
      class="ddm-label"
      tabindex="0"
    >
      label
    </label>
    <input
      class="form-control ddm-field-text"
      id="textField"
      name="textField"
      type="text"
      value=""
    />
    <input
      name="textField_edited"
      type="hidden"
      value="false"
    />
    <span
      class="sr-only"
      id="textField_fieldDetails"
    >
      label
    </span>
  </div>
</div>
`;

exports[`Field Text has a placeholder 1`] = `
<div>
  <div
    aria-labelledby="textField_fieldDetails"
    class="form-group hide"
    data-field-name="textField"
    tabindex="0"
  >
    <input
      class="form-control ddm-field-text"
      id="textField"
      name="textField"
      placeholder="Placeholder"
      type="text"
      value=""
    />
    <input
      name="textField_edited"
      type="hidden"
      value="false"
    />
    
  </div>
</div>
`;

exports[`Field Text has a value 1`] = `
<div>
  <div
    aria-labelledby="textField_fieldDetails"
    class="form-group hide"
    data-field-name="textField"
    tabindex="0"
  >
    <input
      class="form-control ddm-field-text"
      id="textField"
      name="textField"
      type="text"
      value="value"
    />
    <input
      name="textField_edited"
      type="hidden"
      value="false"
    />
    
  </div>
</div>
`;

exports[`Field Text has an id 1`] = `
<div>
  <div
    aria-labelledby="Id_fieldDetails"
    class="form-group hide"
    data-field-name="textField"
    tabindex="0"
  >
    <input
      class="form-control ddm-field-text"
      id="Id"
      name="textField"
      type="text"
      value=""
    />
    <input
      name="textField_edited"
      type="hidden"
      value="false"
    />
    
  </div>
</div>
`;

exports[`Field Text is not readOnly 1`] = `
<div>
  <div
    aria-labelledby="textField_fieldDetails"
    class="form-group hide"
    data-field-name="textField"
    tabindex="0"
  >
    <input
      class="form-control ddm-field-text"
      id="textField"
      name="textField"
      type="text"
      value=""
    />
    <input
      name="textField_edited"
      type="hidden"
      value="false"
    />
    
  </div>
</div>
`;

exports[`Field Text is not required 1`] = `
<div>
  <div
    aria-labelledby="textField_fieldDetails"
    class="form-group hide"
    data-field-name="textField"
    tabindex="0"
  >
    <input
      class="form-control ddm-field-text"
      id="textField"
      name="textField"
      type="text"
      value=""
    />
    <input
      name="textField_edited"
      type="hidden"
      value="false"
    />
    
  </div>
</div>
`;

exports[`Field Text is readOnly 1`] = `
<div>
  <div
    aria-labelledby="textField_fieldDetails"
    class="form-group hide"
    data-field-name="textField"
    tabindex="0"
  >
    <input
      class="form-control ddm-field-text"
      disabled=""
      id="textField"
      name="textField"
      type="text"
      value=""
    />
    <input
      name="textField_edited"
      type="hidden"
      value="false"
    />
    
  </div>
</div>
`;

exports[`Field Text renders Label if showLabel is true 1`] = `
<div>
  <div
    class="form-group hide"
    data-field-name="textField"
  >
    <label
      aria-labelledby="textField_fieldDetails"
      class="ddm-label"
      tabindex="0"
    >
      text
    </label>
    <input
      class="form-control ddm-field-text"
      id="textField"
      name="textField"
      type="text"
      value=""
    />
    <input
      name="textField_edited"
      type="hidden"
      value="false"
    />
    <span
      class="sr-only"
      id="textField_fieldDetails"
    >
      text
    </span>
  </div>
</div>
`;
